src/FsAutoComplete/LspServers/ProjectWorkspace.fs

```diff
@@ -0,0 +1,298 @@
+          (createReferences cachedSnapshots inMemorySourceFiles sourceTextFactory loadedProjectsA)
+        |> createReferencedProjectsFSharpReference resolvedTargetPath
+      else
+        // TODO: Find if this needs to be adaptive or if `getStamp` in a PEReference will be enough to break thru the caching in FCS
+        loadFromDotnetDll proj |> AVal.constant)
+    |> AMap.toASetValues
+
+  and optionsToSnapshot
+    (cachedSnapshots: Dictionary<_, _>)
+    (inMemorySourceFiles: amap<_, aval<VolatileFile>>)
+    (sourceTextFactory: aval<ISourceTextFactory>)
+    (mapReferences: ProjectOptions -> aset<aval<FSharpReferencedProjectSnapshot>>)
+    (p: ProjectOptions)
+    =
+    let normPath = Utils.normalizePath p.ProjectFileName
+
+    match cachedSnapshots.TryGetValue normPath with
+    | true, snapshot ->
+      logger.debug (
+        Log.setMessage "optionsToSnapshot - Cache hit - {projectFileName}"
+        >> Log.addContextDestructured "projectFileName" p.ProjectFileName
+      )
+
+      snapshot
+    | _ ->
+      logger.debug (
+        Log.setMessage "optionsToSnapshot - Cache miss - {projectFileName}"
+        >> Log.addContextDestructured "projectFileName" p.ProjectFileName
+      )
+
+      let projectName = AVal.constant p.ProjectFileName
+      let projectId = AVal.constant p.ProjectId
+
+
+      let sourceFiles = // alist because order matters for the F# Compiler
+        p.SourceFiles
+        |> AList.ofList
+        |> AList.map Utils.normalizePath
+        |> AList.map (fun sourcePath ->
+
+          aval {
+            match! inMemorySourceFiles |> AMap.tryFind sourcePath with
+            | Some volatileFile -> return! volatileFile |> AVal.map createFSharpFileSnapshotInMemory
+            | None -> return! createFSharpFileSnapshotOnDisk sourceTextFactory sourcePath
+          })
+
+      let references = p.OtherOptions |> List.filter (fun x -> x.StartsWith("-r:"))
+
+      let otherOptions = p.OtherOptions |> ASet.ofList |> ASet.map (AVal.constant)
+
+      let referencePaths =
+        references
+        |> ASet.ofList
+        |> ASet.map (fun referencePath ->
+          referencePath.Substring(3) // remove "-r:"
+          |> createReferenceOnDisk)
+
+      let referencedProjects = mapReferences p
+      let isIncompleteTypeCheckEnvironment = AVal.constant false
+      let useScriptResolutionRules = AVal.constant false
+      let loadTime = AVal.constant p.LoadTime
+      let unresolvedReferences = AVal.constant None
+      let originalLoadReferences = AVal.constant []
+
+      let snap =
+        makeAdaptiveFCSSnapshot2
+          projectName
+          projectId
+          sourceFiles
+          referencePaths
+          otherOptions
+          referencedProjects
+          isIncompleteTypeCheckEnvironment
+          useScriptResolutionRules
+          loadTime
+          unresolvedReferences
+          originalLoadReferences
+
+      cachedSnapshots.Add(normPath, snap)
+
+      snap
+
+  let createSnapshots
```

Comment:
TODO: Comments for this

This allows us to create the DAG of snapshots. Various changes to parent snapshots (Source file changes, referenced assembly changes) will propagate creating new snapshots to its children. This type of stuff used to be handled by the compiler but now is forced to happen externally.